packages/ui-react-utils/src/DeterministicIdContext/withDeterministicId.tsx

```diff
@@ -0,0 +1,110 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015 - present Instructure, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+import React, {
+  forwardRef,
+  PropsWithoutRef,
+  RefAttributes,
+  useContext
+} from 'react'
+import hoistNonReactStatics from 'hoist-non-react-statics'
+
+import { DeterministicIdContext } from './DeterministicIdContext'
+import { decorator } from '@instructure/ui-decorator'
+import { generateId } from '@instructure/ui-utils'
+
+import type { DeterministicIdProviderValue } from './DeterministicIdProvider'
+import type { InstUIComponent } from '@instructure/shared-types'
+import { warn } from '@instructure/console'
+
+type WithDeterministicIdProps = {
+  instanceMapCounter?: DeterministicIdProviderValue
+  deterministicId?: string
+}
```

Comment:
Yes, technically they are always there (since the decorator provides them), but if I change it to non-optional, then on the user side it will complain if the user does not provide those props to the component.
That is why it would be a good idea to have 2 separate props type, one for our internal usage and one for the user facing props.

Reply:
Oooooh you are right, forgot about that! Nvm then :) Having to props type sound interesting, although I'm not sure how that would work.. might be worth investigating.